src/inc/Lfo.h

```diff
@@ -0,0 +1,88 @@
+#if !defined(__Lfo_hdr__)
+#define __Lfo_hdr__
+
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include "ErrorDef.h"
+#include "Synthesis.h"
+#include "RingBuffer.h"
+
+class CLfo
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
+    {
+        m_pCRingBuffer = 0;
+        m_fRead = 0;
+        m_fModFreq = fModFreq;
+        m_fSampleRate = fSampleRate;
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        m_pCRingBuffer = new CRingBuffer<float>(m_iBufferLength);
+    }
+    
+    ~CLfo()
+    {
+        delete m_pCRingBuffer;
+        m_pCRingBuffer = 0;
+    }
+    
+    Error_t setParam(float fParamValue)
+    {
+        m_fModFreq = fParamValue;
+        float *pSine = new float [m_iBufferLength];
+        
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        }
+        else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        delete [] pSine;
+        
+        return Error_t::kNoError;
+    }
+    
+    //return current values
+    float returnLfoVal()
+    {
+        float fLfo = m_pCRingBuffer->get(m_fRead);
+        
+        m_fRead = m_fRead + 1;
+        
+        if (m_fRead >= m_iBufferLength) {
+            m_fRead -= m_iBufferLength;
+        }
+        
+        return fLfo;
+        
+    }
+    
+    void process()
+    {
+        float *pSine = new float [m_iBufferLength];
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        
+        delete [] pSine;
+    }
```

Comment:
Having memory allocation and deletion and generating a sinusoid "m_iBufferLength" long each time the process called may not be the most optimal. Also, you could definitely condense your `float returnLfoVal()` function and your `void process()` function into a single `float process()` function

Reply:
Thanks for advising, now it's improved!
```
Vibrato.h - Line 118
Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {
    float fModAmp, fDelay = 0;
    for(int i = 0; i < m_iNumChannels; i++) {
        for(int j = 0; j < iNumberOfFrames; j++) {
            fModAmp = m_pCLfo -> getValue();
            fDelay = 1 + m_fDelayInS * m_fSampleRateInHz + m_fModWidthInS * m_fSampleRateInHz * fModAmp;
            m_ppCRingBuffer[i] -> putPostInc(ppfInputBuffer[i][j]);
            ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
            m_ppCRingBuffer[i] -> getPostInc();
        }
    }
    return Error_t::kNoError;
}
```
```
Lfo - Line 65
    float getValue () {
        return m_pRingBuffer -> getPostInc();
    }
```